dask/dataframe/shuffle.py

```diff
@@ -300,3 +243,244 @@ def collect(group, p, meta, barrier_token):
+        ind = df[index]
+    else:
+        ind = index
+    parts = pd.Series(divisions).searchsorted(ind, side='right') - 1
+    parts[(ind == divisions[-1]).values] = len(divisions) - 2
+    result = (df.assign(partitions=parts, new_index=ind)
+                .set_index('partitions', drop=drop))
+    return result
+
+
+def shuffle_pre_partition_scalar(df, index, divisions, drop):
+    ind = df[index]
+    parts = pd.Series(divisions).searchsorted(ind, side='right') - 1
+    parts[(ind == divisions[-1]).values] = len(divisions) - 2
+    result = (df.assign(partitions=parts)
+                .set_index('partitions', drop=drop))
+    return result
+
+
+def shuffle_pre_partition_series(df, index, divisions, drop):
+    parts = pd.Series(divisions).searchsorted(index, side='right') - 1
+    parts[(index == divisions[-1]).values] = len(divisions) - 2
+    result = (df.assign(partitions=parts, new_index=index)
+                .set_index('partitions', drop=drop))
+    return result
+
+
+def shuffle_group(df, stage, k):
+    if pd.__version__ >= '0.17':
+        index = df.index // k ** stage % k
+    else:
+        values = df.index.values // k ** stage % k
+        index = df.index.copy()
+        index.values[:] = values
+    inds = set(index.drop_duplicates())
+    df = df.set_index(index)
+
+    result = dict(((i, df.loc[i] if i in inds else df.head(0))
+                  for i in range(k)))
+    if isinstance(df, pd.DataFrame):
+        result = dict((k, pd.DataFrame(v).transpose()
+                           if isinstance(v, pd.Series) else v)
+                        for k, v in result.items())
+
+    return result
+
+
+def shuffle_post_scalar(df, index_name):
+    return df.set_index(index_name, drop=True)
+
+
+def shuffle_post_series(df, index_name):
+    df = df.set_index('new_index', drop=True)
+    df.index.name = index_name
+    return df
+
+
+def set_partition_tasks(df, index, divisions, max_branch=32, drop=True,
+                        token=None):
+    max_branch = max_branch or 32
+    n = df.npartitions
+    assert len(divisions) == n + 1
+    name = 'shuffle'
+
+    stages = int(math.ceil(math.log(n) / math.log(max_branch)))
+    if stages > 1:
+        k = int(math.ceil(n ** (1 / stages)))
+    else:
+        k = n
+
+    groups = []
+    splits = []
+    joins = []
+
+    inputs = [tuple(digit(i, j, k) for j in range(stages))
+              for i in range(n)]
+
+    sinputs = set(inputs)
+
+    if np.isscalar(index):
+        meta = shuffle_pre_partition_scalar(df._pd, index, divisions, drop)
+        meta = meta.drop(index, axis=1)
+        df2 = map_partitions(shuffle_pre_partition_scalar, meta, df, index,
+                             divisions, drop)
+    else:
+        meta = df._pd.copy()
+        meta.index = index._pd
+        df2 = map_partitions(shuffle_pre_partition_series, meta, df, index,
+                             divisions, drop)
+
+    token = token or tokenize(df, index, divisions, max_branch)
```

Comment:
Either needs `drop`, or `df` should be `df2` (and you could drop all but `df2` and `max_branch` then).